[wip] bundle ssbd

av8ta opened the pull request:
Hi @RangerMauve here's an initial implementation of ssb bundled with agregore. I've written `ssbd` as a much simpler alternative to shipyard. It's an ssb server with flexible plugin loading; you basically pass an array of plugins to load. 

I've included the list of plugins from https://github.com/metacentre/shipyard-ssb/blob/master/index.js to start with. This list gives you the same sbot as `ssb-server`, which is based on `ssb-db`. 

We can of course develop a default list for agregore which gives us the option of using `ssb-db2` (`ssb-fetch` already works with either `ssb-db` or `ssb-db2`). Plugins to load are configurable by the `.agregorerc` file so the user is in control. 

Figured it would be a good idea to put a simple implementation in place and raise this pull request so we can discuss what the ux might be.

av8ta commented:
I'm thinking since we already have urls like `agregore://welcome`, why not have protocol specific extensions mounted at paths appended to the protocol string? Perhaps an extension at `ssb://settings` could be directed to when onboarding to ssb. The user could then set their existing ssb server config, or could choose to use the defaults provided by agregore for a new ssb member. 

If you're open to something like this I'd like to implement that first before turning to the question of invites to an agregore pub or perhaps an agregore ssb-room for new users to receive content immediately.

RangerMauve commented:
> 'm thinking since we already have urls like agregore://welcome, why not have protocol specific extensions mounted at paths appended to the protocol string? Perhaps an extension at ssb://settings could be directed to when onboarding to ssb.

Would you mind opening a separate issue to talk about that? It feels like this use case would be better suited for [navigator.registerprotocolhandler](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler)

RangerMauve commented:
Gonna review the SSB specific stuff later tonight 🎉

RangerMauve commented on app/protocols/ssb-protocol.js:
```diff
@@ -3,9 +3,20 @@ const fetchToHandler = require('./fetch-to-handler')
 
-    const fetch = makeSsbFetch(options)
+    /** connect to running ssb-server */
+    if (!!options.ssbd === false) {
```
Would it make sense to try to do an HTTP request to a local sbot server to see if it exists before initializing ssbd?

av8ta replied:
Not all sbot's have an http server. They're muxrpc based. Some do have an http server if they're running `ssb-ws` or have added a server, but it's not required. That's if I'm understanding your request correctly :)

[`ssb-party`](https://www.npmjs.com/package/ssb-party) has the functionality you want but hasn't been maintained for quite some time I think. It's certainly not a big deal to implement this and I had considered it, but I find myself that I sometimes have a couple of different sbots running at the same time on different ports and it's super easy to connect to the one you want by setting the `ssb_appname` env var to the server you want.  

That functionality is already implemented in `ssb-fetch`. The way it works with this pull request is:

* default config baked into agregore - `~/.agregore-ssb`
* user config that overrides in `~/.agregorerc`
* env var that overrides the others e.g. `export ssb_appname=ssb`

So if you're a new member of the scuttleverse you can just use agregore to access the network without having to set anything.

If you're an existing user and have a log in `~/.ssb` you can set `appname` to `ssb` to use that identity. You can either use the bundled `ssbd` server or set `ssbd.runServer` to false to connect to a running sbot.

And then if you're hacking on ssb and running more than one sbot it's easy to manage with `ssb_appname` env var.

Different configurations are possible by twiddling the env var and `.agregorerc` for flexibility and it all works as you'd expect (much better than my earlier attempt with shipyard!). And in the simple case you don't have to configure anything. Just run agregore.

RangerMauve commented on app/config.js:
```diff
@@ -56,8 +56,16 @@ module.exports = require('rc')('agregore', {
+  ssbOptions: {
+    appname: DEFAULT_SSB_APPNAME,
+    ssbd: {
+      plugins: require('@metacentre/shipyard-ssb')
```
Since these settings need to be JSON-compatible, could we have the plugins be an array of strings instead and do the require within ssb-protocol.js?

```suggestion
      plugins: ['@metacentre/shipyard-ssb']
```

av8ta replied:
The node resolution algorithm makes this painful to do. I was doing exactly this (loading from bare specifier strings) in shipyard and found while it did work the user experience wasn't great. Kinda violated the principle of least surprise. 

So long as I installed all my ssb plugins in `~/node_modules` life was good. That's why I initially wrote ssbd to only consume already require'd modules. But of course `.agregorerc` is json so I had to put the requiring of strings into ssbd. 

In my opinion, since agregore is bundling an ssb daemon, the simplest thing is to require plugins in agregore for the default case of new users. 

For users who are already on ssb and want to choose their own list of ssb plugins, they can do that in `.agregorerc`:

```json
"ssbOptions": {
    "ssbd": {
      "plugins": ["/absolute/path/to/list/of/plugins/to/require"]
    }
  }
```

Clearly I need to document this!

Setting the `.agregorerc` file could be done from `agregore://settings/ssb` for a smooth user experience.

RangerMauve reviewed (changes requested):
Looking great so far, ty!

Would it also be possible to add some docs to the README about how this works?

RangerMauve commented:
🤔 A browser UI for configuring settings in general like `agregore://settings` might be useful. Kinda like how chrome://flags works.

av8ta commented:
> thinking A browser UI for configuring settings in general like agregore://settings might be useful. Kinda like how chrome://flags works.

yeah, that's what I was thinking, except why not have that for each protocol as well? They are different concerns. But the advantage of doing it under `agregore://settings` would be that we don't clobber any of the host portions of protocol urls. It's not a problem with `ssb://` because the urls are all base64-ish hashes... however if we were to implement an ssb dns-like system... okay I've talked myself into centralising settings under `agregore://settings` while typing :rofl: 

How about we have one extension bundled with agregore at `agregore://settings` where general settings for agregore would go, as well as a route for each protocol under that? `agregore://settings/ssb` etc? Each protocol handler could either have its settings ui hardcoded in that extension, or perhaps be able to register it with the extension? So that way each protocol handler can independently update its settings ui without needing to make a pull request to agregore. What do you think?

av8ta commented on app/protocols/ssb-protocol.js:
```diff
@@ -6,16 +6,14 @@ module.exports = async function createHandler (options, session) {
 
     /** connect to running ssb-server */
-    if (!!options.ssbd === false) {
+    if (!!options.ssbd?.runServer === false) {
```
> Would it also be possible to add some docs to the README about how this works?

I've changed this to connect to an already running sbot if ssbOptions.ssbd.runServer is false and documented it in the docs folder:

https://github.com/AgregoreWeb/agregore-browser/pull/196/commits/b2af81501c8b092aef08974ad3b28a51cb51fdc9#diff-f89bb456646ffb1a9491c3a422072f31b8320e29d00205ed4d3b7be04dbcd660R28

av8ta commented:
> thinking A browser UI for configuring settings in general like `agregore://settings` might be useful. Kinda like how chrome://flags works.

Raised an issue to discuss this one: https://github.com/AgregoreWeb/agregore-browser/issues/198

av8ta commented:
Converted this to draft @RangerMauve because I want to add tests to ssbd first.

Related; what are your thoughts on e2e / regression / integration testing? Would be reassuring if we could run tests to ensure we haven't broken something!

av8ta commented:
Added tests to ssbd.

Added config scenarios to docs: [dd4ba4c](https://github.com/AgregoreWeb/agregore-browser/pull/196/commits/dd4ba4ca91bec1c5c1fecc65b92ce3aedd2eea62) and manually tested the scenarios work as expected.

RangerMauve commented on app/config.js:
```diff
@@ -56,8 +56,17 @@ module.exports = require('rc')('agregore', {
+    appname: DEFAULT_SSB_APPNAME,
+    ssbd: {
+      runServer: true,
+      plugins: require('@metacentre/shipyard-ssb')
```
Yeah, still feels like this should be a list of paths to require plugins from. 🤔

e.g. if you use `npm install` in some other folder, you could still add the plugin path inside your local node_modules.

Like

```
cd ~/
npm install my-plugin
```

Then have a config that looks like

{
  ssbOptions: {
    ssbd: {
      plugins: ["@metacentre/shipyard-ssb", "~/node_modules/my-plugin"]
    }
  }
}

RangerMauve replied:
Whoops, totally missed the part of the docs which covered that you already implemented this. 😅

RangerMauve commented on docs/Fetch-SSB.md:
````diff
@@ -0,0 +1,198 @@
+  },
+```
+
+The ssb data directory resides at `~/.agregore-ssb`
````
Is there a way we could place the data inside a `ssb` folder within the Agregore data directory beside everything else?

av8ta replied:
That was the first thing I tried and it was HARD. ssb really expects data in the home directory even though it uses rc under the hood. I think perhaps the ssb-config module is opinionated on this. I could have another go, but it may be a hair tearing exercise haha

RangerMauve replied:
That's okay we can just keep it as is for the first iteration. 🤷

RangerMauve commented on app/config.js:
```diff
@@ -56,8 +56,17 @@ module.exports = require('rc')('agregore', {
+    appname: DEFAULT_SSB_APPNAME,
+    ssbd: {
+      runServer: true,
+      plugins: require('@metacentre/shipyard-ssb')
```
If we can require absolute paths, we can probably require regular paths too, right?

```suggestion
      plugins: ['@metacentre/shipyard-ssb']
```

av8ta replied:
This was ages ago but my brain says something-something-something-node-resolution-algorithm. I did this with shipyard and found it a nightmare. Honestly, it was much smoother just requiring a module that required the plugins you want.

RangerMauve replied:
Cool, I'll use `require.resolve` instead then if that's okay

RangerMauve replied:
Or honestly we can keep it for now. 🤷

RangerMauve commented:
TY for your patience with this! Gonna release it along with some fixes in a sec. 😁

RangerMauve commented:
After merging this I'm getting errors with `nj-cli` being missing when I try to install deps.

Specifically looks like this:

```
rror /home/mauve/programming/agregore-browser/node_modules/ssb-validate2-rsjs-node: Command failed.
Exit code: 1
Command: node install.js
Arguments: 
Directory: /home/mauve/programming/agregore-browser/node_modules/ssb-validate2-rsjs-node
Output:
node:internal/errors:477
    ErrorCaptureStackTrace(err);
    ^

Error: spawn nj-cli ENOENT
    at ChildProcess._handle.onexit (node:internal/child_process:283:19)
    at onErrorNT (node:internal/child_process:476:16)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'spawn nj-cli',
  path: 'nj-cli',
  spawnargs: [ 'build', '--release' ]
```

RangerMauve commented:
This seems to be related to neon bindings or something?

I'm afraid I'll need to revert this since I can't get the builds to work. 😭

av8ta commented:
Hey thanks for coming back to this. Sorry it's not building! Are you building on apple m1 silicon? I saw that error on `ssb-validate2-rsjs-node` when building on a pi a few months back. From what I recall it was a rust toolchain issue. I'll look into it tomorrow :)

RangerMauve commented:
I was building on x64 Intel Linux actually. Pop OS based on Ubuntu 21
